rust_src/src/keyboard.rs

```diff
@@ -122,6 +123,36 @@ pub fn lucid_event_type_list_p(event: Option<LispCons>) -> bool {
     })
 }
 
+// _Noreturn will be added to prototype by make-docfile.
```

Comment:
What does this mean and how do i add attributes to functions ?

Reply:
According to @db48x, there's no need to worry about it in Rust (see https://github.com/Wilfred/remacs/pull/904#issuecomment-427613854). Here, it's taken care of by the `!` return type. In C, the generated prototype becomes `_Noreturn EXFUN (Fabort_recursive_edit, 0);`, whatever that means.

Reply:
From the GCC docs:

```
noreturn
A few standard library functions, such as abort and exit, cannot return. GCC knows this automatically. Some programs define their own functions that never return. You can declare them noreturn to tell the compiler this fact. For example,
          void fatal () __attribute__ ((noreturn));
          
          void
          fatal (/* ... */)
          {
            /* ... */ /* Print error message. */ /* ... */
            exit (1);
          }
     
The noreturn keyword tells the compiler to assume that fatal cannot return. It can then optimize without regard to what would happen if fatal ever did return. This makes slightly better code. More importantly, it helps avoid spurious warnings of uninitialized variables.

The noreturn keyword does not affect the exceptional path when that applies: a noreturn-marked function may still return to the caller by throwing an exception or calling longjmp.

Do not assume that registers saved by the calling function are restored before calling the noreturn function.

It does not make sense for a noreturn function to have a return type other than void.

The attribute noreturn is not implemented in GCC versions earlier than 2.5. An alternative way to declare that a function does not return, which works in the current version and in some older versions, is as follows:

          typedef void voidfn ();
          
          volatile voidfn fatal;
```

Reply:
TL;DR, it allows for compilation optimization and checks in C code. It informs how we port using `-> !` syntax as @nickdrozd points out.

Reply:
We can drop the comment